### PR TITLE
[compiler] Initialize abstract values of places in InferReferenceEffects

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30976
* #30975
* #30974
* __->__ #30973
* #30972

Test Plan:
In this diff, we now populate the abstractValue field of places during the InferReferenceEffects pass. The value we populate it with is the same value that we use internally in this pass, but it now will remain accessible to downstream phases as part of the Place.

For phis specifically, we need to do a bit of extra work to compute the appropriate value for the phi, since InferReferenceEffects currently doesn't need to infer a value for phis themselves, only values downstream of them. However, the value we compute should correspond to the value available downstream of the phi.

This changes the error message for one todo test case, because we now are querying for the valueKind earlier in the pass, and hitting an invariant violation as a result of that rather than a later invariant violation.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts [hunk 1]
```diff
@@ -299,11 +305,11 @@ class InferenceState {
...
 
   // Lookup the kind of the given @param value.
-  kind(place: Place): AbstractValue {
+  kind(place: {identifier: Identifier; loc: SourceLocation}): AbstractValue {
     const values = this.#variables.get(place.identifier.id);
     CompilerError.invariant(values != null, {
-      reason: `[hoisting] Expected value kind to be initialized`,
-      description: `${printPlace(place)}`,
+      reason: `[hoisting] Expected value kind to be initialized in call to kind()`,
+      description: `${place.identifier.id}`,
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts:312]: Nit: printing the entire identifier / place might be more descriptive (or is this change intentional?)

**Review by josephsavona — APPROVED** [review]
